lib/jobs/base.ex

```diff
@@ -0,0 +1,175 @@
+      @job_name Keyword.fetch!(opts, :name)
+      @repo Keyword.fetch!(opts, :repo)
+      # @behaviour behaviour
+      require Logger
+
+      def log_for_job(message, level \\ :info) do
+        %{jid: exq_job_id} = Exq.worker_job()
+        message = "#{__MODULE__}[#{exq_job_id}]: #{message}"
+
+        case level do
+          :debug -> Logger.debug(message)
+          :error -> Logger.error(message)
+          :info -> Logger.info(message)
+          :warn -> Logger.warn(message)
+          _ -> Logger.info(message)
+        end
+      end
+
+      def get_active_execution_lock_for_hash(hash) do
+        Model
+        |> where([m], m.hash == ^hash)
+        |> where([m], is_nil(m.completed_at))
+        |> first()
+        |> unquote(@repo).one()
+      end
+
+      def get_latest_execution_for_hash(hash) do
+        Model
+        |> where([m], m.hash == ^hash)
+        |> order_by(desc: :completed_at)
+        |> first()
+        |> unquote(@repo).one()
+      end
+
+      def insert_execution_lock(hash, encoded_params, attempt) do
+        unquote(@repo).insert!(%Model{
+          attempt: attempt,
+          hash: hash,
+          encoded_parameters: encoded_params,
+          job_name: unquote(@job_name)
+        })
+      end
+
+      def update_execution(execution_record, attrs \\ %{}) do
+        execution_record
+        |> Model.changeset(attrs)
+        |> unquote(@repo).update!()
+      end
+
+      def get_hash(locked_params) do
+        param_binary =
+          locked_params
+          |> Map.put(:job_name, unquote(@job_name))
+          |> Map.to_list()
+          |> Enum.sort()
+          |> :erlang.term_to_binary()
+
+        hash = :crypto.hash(:sha256, param_binary) |> Base.encode64()
+
+        {:ok, hash}
+      end
+
+      def encode_params(params) do
+        params
+        |> Map.to_list()
+        |> Enum.sort()
+        |> :erlang.term_to_binary()
+        |> Base.encode64()
+      end
+
+      def execution_locked_with_param_hash(params) do
+        with locked_params <- lock_params(params),
+             false <- is_nil(locked_params) || Enum.empty?(locked_params),
+             {:ok, hash} <- get_hash(locked_params) do
+          active_execution = get_active_execution_lock_for_hash(hash)
+          {active_execution, hash}
+        else
+          true ->
+            log_for_job("#{__MODULE__}.lock_params/1 did not return a lockable state, skipping to execute/1.")
+            {false, nil}
+        end
+      end
+
+      def perform(params) do
+        try do
+          atomized_params = params |> RequestHelpers.recursively_atomize()
+          log_for_job("determining lock state.")
+          {locker_record, hash} = execution_locked_with_param_hash(atomized_params)
+
+          if locker_record, do: raise(:locked)
+
+          previous_execution = get_latest_execution_for_hash(hash)
+
+          attempt_number = if previous_execution != nil, do: Map.get(previous_execution, :attempt) + 1, else: 1
```

Comment:
Is it not better to let exq handle this info?  Why track the attempt manually?

Reply:
Because an attempt in EXQ is for a specific job id, in our case it is a hash of the lock. let's chat